rx_signal_processing/rx_dsp_chain.cu

```diff
@@ -15,406 +15,238 @@
-                                                                                 //REPLY could be. All values used at this point are uint32 so it should follow truncation rules.
-        dp->allocate_second_stage_output(num_output_samples_2);
-        auto samples_per_antenna_2 = cp.numberofreceivesamples()/first_stage_dm_rate; // REVIEW #26 num_samps_2 could be better named, samples2_per_antenna ?
-                                                                            // REPLY okay
-        dp->call_decimate(dp->get_first_stage_output_p(),
-            dp->get_second_stage_output_p(),
-            dp->get_second_stage_filters_p(), second_stage_dm_rate,
-            samples_per_antenna_2, filtertaps_2.size(), rx_freqs.size(),
-            total_antennas, "Second stage of decimation"); // REVIEW #0 first_stage_output_p points to array of data that has separate datapoints for different frequencies, is this taken into account?
-// REVIEW #0 should we pass rx_freqs.size() or should we just pass = 1 because there is only one lowpass filter in this stage? Perhaps should write another decimate function for this.
-//REPLY might need a redesign here
-
-// REVIEW #0 current call_decimate function is built for the bandpass filters (to separate data by frequency) - for the second stage, we don't need this so a separate function would be better
-// REPLY Might need a redesign here, but perhaps not. I find it more advantageous to minimize the amount of cuda to have to debug
-
-        dp->allocate_and_copy_third_stage_filters(filtertaps_3_h.data(), filtertaps_3_h.size());
-        auto num_output_samples_3 = num_output_samples_2 / third_stage_dm_rate;
-        dp->allocate_third_stage_output(num_output_samples_3);
-        auto num_samps_3 = num_samps_2/second_stage_dm_rate;
-        dp->call_decimate(dp->get_second_stage_output_p(),
-            dp->get_third_stage_output_p(),
-            dp->get_third_stage_filters_p(), third_stage_dm_rate,
-            num_samps_3, filtertaps_3.size(), rx_freqs.size(),
-            total_antennas, "Third stage of decimation");
-
-        dp->allocate_and_copy_host_output(num_output_samples_3);
-
-        // New in CUDA 5.0: Add a CPU callback which is called once all currently pending operations in the CUDA stream have finished
-        gpuErrchk(cudaStreamAddCallback(dp->get_cuda_stream(),
-                                            DSPCore::cuda_postprocessing_callback, dp, 0));
+    // each antenna has a data set for each frequency after filtering.
+    auto samples_per_antenna_2 = rx_freqs.size() *
+                                  rx_metadata.numberofreceivesamples()/first_stage_dm_rate;
+    call_decimate<DecimationType::lowpass>(dp->get_first_stage_output_p(),
+      dp->get_second_stage_output_p(),
+      dp->get_second_stage_filter_p(), second_stage_dm_rate,
+      samples_per_antenna_2, filters.get_second_stage_lowpass_taps().size(), rx_freqs.size(),
```

Comment:
samples_per_antenna_2 Could be simplified into num_output_samples_1 / total_antennas